### PR TITLE
feat: worker support

### LINKED ISSUE #3074: WorkerDependency

### PR DESCRIPTION
**Pull request by ahabhgk**
## Summary

resolve #3074

checkout `example/worker` for example

Feature:
- output.workerChunkLoading
- output.workerWasmLoading
- output.workerPublicPath
- worker with `output.module: true`

Todo:
- webpackEntryOptions magic comments (low priority)
- https://github.com/web-infra-dev/rspack/issues/3656 (low priority, block by https://github.com/web-infra-dev/rspack/discussions/3619)
- https://github.com/web-infra-dev/rspack/issues/3658 **not work with tree shaking, will fix it in next PR**

## Test Plan

`example/worker` works, and some tests block by #3658, #3655, #2408, will add more tests after blockers resolved

### CODE REVIEW

**Comment by IWANABETHATGUY** [comment]
!bench

File: crates/node_binding/binding.d.ts [hunk 1]
```diff
@@ -371,6 +371,12 @@ export interface JsStatsWarning {
   formatted: string
 }
 
+export interface NodeFS {
+  writeFile: (...args: any[]) => any
+  mkdir: (...args: any[]) => any
```
Comment (underfin) [inline crates/node_binding/binding.d.ts:376]: @h-a-n-a cc

File: crates/rspack_plugin_javascript/src/visitors/dependency/url_scanner.rs [hunk 2]
```diff
@@ -7,50 +7,58 @@ use swc_core::ecma::{
...
+
+pub fn match_new_url(new_expr: &NewExpr) -> Option<(u32, u32, String)> {
+  if matches!(&*new_expr.callee, Expr::Ident(Ident { sym: js_word!("URL"), .. }))
+  && let Some(args) = &new_expr.args
+  && let (Some(first), Some(second)) = (args.first(), args.get(1))
+  && let (
+    ExprOrSpread { spread: None, expr: box Expr::Lit(Lit::Str(path)) },
+    ExprOrSpread { spread: None, expr: box expr },
+  ) = (first, second) && expr_matcher::is_import_meta_url(expr) {
+    return Some((path.span.real_lo(), expr.span().real_hi(), path.value.to_string()))
```
Comment (IWANABETHATGUY) [inline crates/rspack_plugin_javascript/src/visitors/dependency/url_scanner.rs:39]:
```suggestion
   if matches!(&*new_expr.callee, Expr::Ident(Ident {sym: js_word!("URL"),..}))
    && let Some(args) = &new_expr.args 
    && let [ExprOrSpread { spread: None, expr: box Expr::Lit(Lit::Str(path)) }, ExprOrSpread { spread: None, expr: box expr }] = &args[..]
    && expr_matcher::is_import_meta_url(expr) {
    return Some((path.span.real_lo(), expr.span().real_hi(), path.value.to_string()))
```

NIT: you can save a let chain

File: crates/rspack_plugin_javascript/src/visitors/dependency/mod.rs [hunk 3]
```diff
@@ -106,7 +112,18 @@ pub fn scan_dependencies(
...
+    program.visit_with(&mut worker_syntax_scanner);
+    let mut worker_scanner = WorkerScanner::new(
+      &module_identifier,
+      &compiler_options.output,
+      worker_syntax_scanner,
+    );
+    program.visit_with(&mut worker_scanner);
+    dependencies.append(&mut worker_scanner.dependencies);
+    presentational_dependencies.append(&mut worker_scanner.presentational_dependencies);
+    program.visit_with(&mut UrlScanner::new(&mut dependencies, &worker_scanner));
```
Comment (underfin) [inline crates/rspack_plugin_javascript/src/visitors/dependency/mod.rs:126]: Look it is bad for me.

File: crates/rspack_plugin_runtime/src/runtime_plugin.rs [hunk 4]
```diff
@@ -39,12 +39,15 @@ impl Plugin for RuntimePlugin {
...
-      .get_all_async_chunks(&compilation.chunk_group_by_ukey)
-      .is_empty()
-      || args
+    if args
+      .runtime_requirements
+      .contains(RuntimeGlobals::ENSURE_CHUNK)
+      && (args
         .runtime_requirements
         .contains(RuntimeGlobals::ENSURE_CHUNK_INCLUDE_ENTRIES)
+        || !chunk
```
Comment (underfin) [inline crates/rspack_plugin_runtime/src/runtime_plugin.rs:48]: Here is wrong.

File: crates/node_binding/binding.d.ts (hunk 1 above)
Comment (h-a-n-a) [reply crates/node_binding/binding.d.ts:376]: This is a problem with NAPI-RS, which should be fixed there

**Review by underfin — DISMISSED** [review]

**Review by underfin — APPROVED** [review]

**Pull request merged**